### PR TITLE
Add a static Bitmap source cache for Inheritance margin

### PR DESCRIPTION
**Pull request by Cosifne**
https://github.com/dotnet/roslyn/issues/54521

As we discussed somewhere offline, this PR change the inheritance margin to our own static image cache instead of the one from the platform.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/CrispImageCache.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.InheritanceMargin.MarginGlyph
+{
+    internal static class CrispImageCache
+    {
+        private static ImageAttributes? s_attributes;
+        private static readonly Dictionary<ImageMoniker, BitmapSource> s_ImageCache = new();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/CrispImageCache.cs:18]: this all feels very non-threadsafe.

File: src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/CrispImageSourceConverter.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        protected override ImageSource Convert(
+            ImageMoniker imageMoniker,
+            double height,
+            double width,
+            double dpi,
+            Color background,
+            object parameter,
+            CultureInfo culture)
+        {
+            var attributes = new ImageAttributes
```
Comment (Cosifne) [inline src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/CrispImageSourceConverter.cs:24]: If we finally want to go this way, some further work needs to be done here to make sure zooming works. (Right now it doesn't zoom based on the zoom level)

File: src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/CrispImageCache.cs (hunk 1 above)
Comment (Cosifne) [reply src/VisualStudio/Core/Def/Implementation/InheritanceMargin/MarginGlyph/CrispImageCache.cs:18]: Sure, I could add a lock to the write part of this dictionaryary

**Comment by Cosifne** [comment]
For me, I didn't see any UI delay after this change using 16.11.
This part of the code is very similar to the [CrispImageConverter](http://index/?query=CrispImageSourceConverter&rightProject=Microsoft.VisualStudio.Imaging&file=Converters%5CCrispImageSourceConverter.cs&line=237). I am feeling the root reason for this is for some reason there is a lot of cache miss [here](http://index/?query=CrispImageSourceConverter&rightProject=Microsoft.VisualStudio.Imaging&file=Converters%5CCrispImageSourceConverter.cs&line=237).

@CyrusNajmabadi  do you want to try it to see if there is still UI delay on your machine?
Also, tag @sharwell for advice.

**Review by sharwell — CHANGES_REQUESTED** [review]
Traces and heap dumps strongly suggest that GC rooting is _not_ the root cause of this problem. I would want a conclusive RCA and an explanation of why the platform's cache cannot handle this case before applying this workaround.

**Pull request closed without merging**